### PR TITLE
Use condy for probes array in Java 11+ class files

### PR DESCRIPTION
**Pull request by Godin**
Users often forget to ignore **synthetic** fields when using reflection and continue asking questions even if this is already described in [our FAQ](https://github.com/jacoco/jacoco/blob/v0.8.3/org.jacoco.doc/docroot/doc/faq.html#L130-L141) - here is some recent ones:

* #834
* #816
* #799
* https://groups.google.com/d/msg/jacoco/3wT1WmrRPBA/A9MPMa6tHAAJ

We can get rid of this field in Java 11+ class files by using condy ([JEP 309](https://openjdk.java.net/jeps/309)).

Also I think that condy might be better in terms of performance than `ClassFieldProbeArrayStrategy` in following cases

* cold non JITed code and I suppose that this is quite common case when running unit tests
* when JIT fails to inline `$jacocoInit`

### CODE REVIEW

File: org.jacoco.core/src/org/jacoco/core/internal/instr/CondyProbeArrayStrategy.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+		this.accessorGenerator = accessorGenerator;
+	}
+
+	public int storeInstance(final MethodVisitor mv, final boolean clinit,
+			final int variable) {
+		final Handle bootstrapMethod = new Handle(Opcodes.H_INVOKESTATIC,
+				className, InstrSupport.INITMETHOD_NAME, B_DESC, isInterface);
+		mv.visitLdcInsn(new ConstantDynamic(InstrSupport.DATAFIELD_NAME,
+				"Ljava/lang/Object;", bootstrapMethod));
+		mv.visitTypeInsn(Opcodes.CHECKCAST, "[Z");
```
Comment (marchof) [inline org.jacoco.core/src/org/jacoco/core/internal/instr/CondyProbeArrayStrategy.java]: Why is CHECKCAST required here? I think I know, but maybe we should also let the reader know ;)

Comment (Godin) [reply org.jacoco.core/src/org/jacoco/core/internal/instr/CondyProbeArrayStrategy.java]: @marchof I moved comment about JDK-8216970 from test into main code

File: org.jacoco.core.test/src/org/jacoco/core/instr/ClassFileVersionsTest.java [hunk 2]
```diff
@@ -147,8 +148,14 @@ public void visitFrame(int type, int nLocal,
 
 							@Override
 							public void visitEnd() {
-								assertEquals(Boolean.valueOf(expected),
-										Boolean.valueOf(frames));
+								if (CondyProbeArrayStrategy.B_DESC
+										.equals(desc)) {
+									assertEquals(Boolean.FALSE,
```
Comment (marchof) [inline org.jacoco.core.test/src/org/jacoco/core/instr/ClassFileVersionsTest.java]: Use `assertFalse()`

Comment (marchof) [reply org.jacoco.core.test/src/org/jacoco/core/instr/ClassFileVersionsTest.java]: Add comment:

    // CondyProbeArrayStrategy does not need frames:

File: org.jacoco.core.test/src/org/jacoco/core/internal/instr/CondyProbeArrayStrategyTest.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.tree.ClassNode;
+import org.objectweb.asm.tree.LdcInsnNode;
+import org.objectweb.asm.tree.MethodNode;
+import org.objectweb.asm.tree.TypeInsnNode;
+import org.objectweb.asm.tree.VarInsnNode;
+
+public class CondyProbeArrayStrategyTest {
+
+	private final CondyProbeArrayStrategy strategy = new CondyProbeArrayStrategy(
```
Comment (marchof) [inline org.jacoco.core.test/src/org/jacoco/core/internal/instr/CondyProbeArrayStrategyTest.java]: Move initialization to `setup()` method for consistency with other tests.

Comment (Godin) [reply org.jacoco.core.test/src/org/jacoco/core/internal/instr/CondyProbeArrayStrategyTest.java]: done

File: org.jacoco.core.test/src/org/jacoco/core/instr/ClassFileVersionsTest.java (hunk 2 above)
Comment (Godin) [reply org.jacoco.core.test/src/org/jacoco/core/instr/ClassFileVersionsTest.java]: > Use `assertFalse()`

done

> Add comment

added as assertion message

**Pull request merged**